File: Java-Core/src/main/java/presto/type/CharacterType.java
package presto.type;

import presto.error.PrestoError;
import presto.error.SyntaxError;
import presto.runtime.Context;
import presto.value.Character;
import presto.value.CharacterRange;
import presto.value.ICollection;
import presto.value.IValue;
import presto.value.Range;
import presto.value.ListValue;

public class CharacterType extends NativeType {

	static CharacterType instance = new CharacterType();
	
	public static CharacterType instance() {
		return instance;
	}
	
	private CharacterType() {
		super("Character");
	}
	
	@Override
	public Class<?> toJavaClass() {
		return java.lang.Character.class;
	}
	

	@Override
	public boolean isAssignableTo(Context context, IType other) {
		return (other instanceof CharacterType) || (other instanceof TextType) || (other instanceof AnyType);
	}

	@Override
	public IType checkAdd(Context context, IType other, boolean tryReverse) throws SyntaxError {
		return TextType.instance();
	}
	
	@Override
	public IType checkMultiply(Context context, IType other, boolean tryReverse) throws SyntaxError {
		if(other instanceof IntegerType)
			return TextType.instance();
		else
			return super.checkMultiply(context, other, tryReverse);
	}
	
	@Override
	public IType checkCompare(Context context, IType other) throws SyntaxError {
		if(other instanceof CharacterType || other instanceof TextType)
			return BooleanType.instance();
		return super.checkCompare(context, other);
	}
	
	@Override
	public IType checkRange(Context context, IType other) throws SyntaxError {
		if(other instanceof CharacterType)
			return new RangeType(this);
		return super.checkRange(context, other);
	}
	
	@Override
	public Range<?> newRange(Object left, Object right) throws SyntaxError {
		if(left instanceof Character && right instanceof Character)
			return new CharacterRange((Character)left,(Character)right);
		return super.newRange(left, right);
	}

	@Override
	public ListValue sort(Context context, ICollection<IValue> list) throws PrestoError {
		return this.doSort(context,list);
	}
	
	@Override
	public String toString(Object value) {
		return "'" + value.toString() + "'";
	}
	
	@Override
	public IValue convertJavaValueToPrestoValue(Object value) {
        if (value instanceof java.lang.Character)
            return new Character(((java.lang.Character)value).charValue());
        else
            return (IValue)value; // TODO for now
	}

}

File: Java-CodeServer/src/main/resources/js/prompto/type/CharacterType.js
var NativeType = require("./NativeType").NativeType;
var BooleanType = require("./BooleanType").BooleanType;
var IntegerType = require("./IntegerType").IntegerType;
var TextType = require("./TextType").TextType;
var AnyType = require("./AnyType").AnyType;
var Character = null;
var RangeType = require("./RangeType").RangeType;
var CharacterRange = require("../value/CharacterRange").CharacterRange;

exports.resolve = function() {
    Character = require("../value/Character").Character;
};

function CharacterType()  {
	NativeType.call(this, "Character");
	return this;
}

CharacterType.prototype = Object.create(NativeType.prototype);
CharacterType.prototype.constructor = CharacterType;

CharacterType.instance = new CharacterType();


CharacterType.prototype.isAssignableTo = function(context, other) {
	return (other instanceof CharacterType) || (other instanceof TextType) || (other instanceof AnyType);
};

CharacterType.prototype.nativeCast = function(context, value) {
    if(value.type instanceof TextType && value.value.length>=1)
        return new Character(value.value.substring(0, 1));
    else
        throw new InvalidDataError("Cannot convert " + value.toString() + " to Character");
};

CharacterType.prototype.checkAdd = function(context, other, tryReverse) {
	return TextType.instance;
};

CharacterType.prototype.checkMultiply = function(context, other, tryReverse) {
	if(other instanceof IntegerType) {
		return TextType.instance;
	}
	return NativeType.prototype.checkMultiply.apply(this, context, other, tryReverse);
};

CharacterType.prototype.checkCompare = function(context, other) {
	if(other instanceof CharacterType || other instanceof TextType) {
		return BooleanType.instance;
	}
	return NativeType.prototype.checkCompare.apply(this, context, other);
};

CharacterType.prototype.checkRange = function(context, other) {
	if(other instanceof CharacterType) {
		return new RangeType(this);
	} else {
		return NativeType.prototype.checkRange.call(this, context, other);
	}
};

CharacterType.prototype.newRange = function(left, right) {
	if(left instanceof Character && right instanceof Character) {
		return new CharacterRange(left, right);
	} else {
		return CharacterType.prototype.newRange.call(this, left, right);
	}
};


exports.CharacterType = CharacterType;